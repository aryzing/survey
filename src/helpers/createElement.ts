const el: typeof document.createElement = (element: string) =>
  document.createElement(element);

export default el;
