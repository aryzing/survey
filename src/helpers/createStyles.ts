const createStyle = (css: string): HTMLStyleElement => {
  const style = document.createElement("style");
  style.textContent = css;

  return style;
};

export default createStyle;
