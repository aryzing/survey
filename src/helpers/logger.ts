/**
 * Simple logger to log errors. Ideally, we'd send them to a monitoring
 * infrastructure, like Sentry.
 */
const logError = (...args: unknown[]): void => {
  // eslint-disable-next-line no-console
  console.error(...args);
};

export default logError;
