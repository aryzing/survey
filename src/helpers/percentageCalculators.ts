import { GQuestion, GAnswer, GStore, GFilter } from "@/store/types";

export const PRa = (answer: GAnswer, question: GQuestion): number => {
  return answer.edgesRespondent.size / question.edgesRespondent.size;
};

export const PRFa = (answer: GAnswer, store: GStore): number => {
  const activeOptions = Array.from(store.Options.values()).filter(
    (o) => o.isActive
  );

  const activeOptionFilterIds = new Set<string>();
  activeOptions.forEach((o) => activeOptionFilterIds.add(o.filterId));

  const activeFilters = Array.from(activeOptionFilterIds.values())
    .map((fId) => store.Filters.get(fId))
    .filter((f) => f) as GFilter[];

  /**
   * Collectino of users filtered according to active options. Respondents
   * are added to this collection when they match any of the active options
   * in a filter, and all filters where an option is active.
   *
   * In other words, options are "OR"ed against the value of a particular
   * demographic, and filters are "AND"ed across all demographics.
   */
  const filteredRespondents = Array.from(store.Respondents.values()).filter(
    (r) => {
      // This reducer acts as the "AND" for filters across all demographics
      return activeFilters.reduce<boolean>(
        (hasPassedPrevFilters, currFilter) => {
          // No need to compute whether the current filter will pass if a
          // previous one has already failed.
          if (!hasPassedPrevFilters) return false;

          const activeOptionsForCurrentFilter = activeOptions.filter(
            (o) => o.filterId === currFilter.id
          );

          // This reducer acts as the "OR" for options in the current filter.
          return activeOptionsForCurrentFilter.reduce<boolean>(
            (hasPassedPrevOptions, currOption) => {
              // No need to compute whether the current option will pass if a
              // previous one has already passed.
              if (hasPassedPrevOptions) return true;

              return r[currFilter.name] === currOption.name;
            },
            false
          );
        },
        true
      );
    }
  );

  const filteredRespondentsThatProvidedCurrentAnswer = filteredRespondents.filter(
    (r) => r.edgesAnswer.has(answer.id)
  );

  return filteredRespondents.length
    ? filteredRespondentsThatProvidedCurrentAnswer.length /
        filteredRespondents.length
    : 0;
};
export const PRo = () => {
  return 3;
};
