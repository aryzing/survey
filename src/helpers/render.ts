import Component from "@/components/Component";
// eslint-disable-next-line import/no-extraneous-dependencies
import { getQueriesForElement, BoundFunctions } from "testing-library__dom";

type BF = BoundFunctions<typeof import("@testing-library/dom/types/queries")>;
type Utils<P, S> = {
  container: HTMLElement;
  component: Component<P, S>;
  cleanup(): void;
};
export type RenderReturn<P, S> = Utils<P, S> & BF;

function render<P2, S2>(
  CustomElement: new (props: P2) => Component<P2, S2>,
  props: P2
): RenderReturn<P2, S2> {
  const div = document.createElement("div");
  const customElement = new CustomElement(props);
  div.append(customElement);
  document.body.append(div);

  const cleanup = () => {
    document.body.removeChild(div);
  };

  const boundFunctions = getQueriesForElement(div);
  const utils = {
    container: div,
    component: customElement,
    cleanup,
  };

  return {
    ...utils,
    ...boundFunctions,
  };
}

export default render;
