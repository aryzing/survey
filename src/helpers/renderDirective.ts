import Component from "@/components/Component";
import { directive, Part, NodePart } from "lit-html";
import logError from "./logger";

const renderFactory = <P, S>(
  CustomElement: new (props: P) => Component<P, S>,
  props: P
) => (part: Part): void => {
  if (!(part instanceof NodePart)) {
    logError(
      "Expected `part` to be instance of `NodePart`. Components can only be rendered into `NodePart`s"
    );
    return;
  }

  if (part.value instanceof CustomElement) {
    part.value.setProps(props);
    return;
  }

  part.setValue(new CustomElement(props));
};

const renderDirective = directive(renderFactory);

export default renderDirective;
