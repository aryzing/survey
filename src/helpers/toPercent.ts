const toRoundedPercent = (n: number): number => {
  return Math.round(n * 100);
};

export default toRoundedPercent;
