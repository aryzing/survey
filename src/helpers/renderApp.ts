import { render, TemplateResult } from "lit-html";
import { logError } from "./logger";

/**
 * Mant to be used only once at the top of the application to render content
 * into the DOM.
 */
// eslint-disable-next-line import/prefer-default-export
export const renderApp = (app: TemplateResult, id: string): void => {
  const rootElement = document.getElementById(id);

  if (!rootElement) {
    logError("Error: could not find root node where to render app.");
  } else {
    render(app, rootElement);
  }
};
