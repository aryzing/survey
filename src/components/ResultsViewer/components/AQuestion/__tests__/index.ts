import { render } from "@/helpers/render";
import type { QuestionWithStats } from "@/components/ResultsViewer/types";
import { AQuestion } from "..";

describe("AQuestion Component", () => {
  test("Renders properly", () => {
    const props: QuestionWithStats = {
      title: "test-question-title",
      question_type: "multiple_choice",
      answers: [
        {
          text: "test-answer-1",
          PRFa: 0.11,
          PRa: 0.12,
        },
        {
          text: "test-answer-2",
          PRFa: 0.21,
          PRa: 0.22,
        },
      ],
      hasActiveFilter: true,
    };
    const { getByText, queryAllByText } = render(AQuestion, props);

    expect(getByText("test-question-title")).toMatchSnapshot();
    expect(getByText("Multiple choice")).toMatchSnapshot();
    expect(getByText("test-answer-1")).toMatchSnapshot();
    expect(queryAllByText("11")).toHaveLength(0);
    expect(getByText("12")).toMatchSnapshot();
    expect(getByText("test-answer-2")).toMatchSnapshot();
    expect(queryAllByText("21")).toHaveLength(0);
    expect(getByText("22")).toMatchSnapshot();
  });
});
