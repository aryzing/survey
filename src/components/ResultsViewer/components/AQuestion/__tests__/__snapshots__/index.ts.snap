// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AQuestion Component Renders properly 1`] = `
<span
  class="question-title"
>
  <!---->
  test-question-title
  <!---->
</span>
`;

exports[`AQuestion Component Renders properly 2`] = `
<span
  class="question-type"
>
  <!---->
  Multiple choice
  <!---->
</span>
`;

exports[`AQuestion Component Renders properly 3`] = `
<span
  class="answer-text"
>
  <!---->
  test-answer-1
  <!---->
</span>
`;

exports[`AQuestion Component Renders properly 4`] = `
<span
  class="percent"
>
  <!---->
  12
  %
</span>
`;

exports[`AQuestion Component Renders properly 5`] = `
<span
  class="answer-text"
>
  <!---->
  test-answer-2
  <!---->
</span>
`;

exports[`AQuestion Component Renders properly 6`] = `
<span
  class="percent"
>
  <!---->
  22
  %
</span>
`;
